backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/predicate/impl/LuceneKnnPredicate.java

```diff
@@ -0,0 +1,119 @@
+import org.apache.lucene.search.Query;
+
+public class LuceneKnnPredicate extends AbstractLuceneSingleFieldPredicate implements LuceneSearchPredicate {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	private final int k;
+	private final Object vector;
+	private final LuceneSearchPredicate filter;
+
+	private LuceneKnnPredicate(Builder<?> builder) {
+		super( builder );
+		this.k = builder.k;
+		this.vector = builder.vector;
+		this.filter = builder.filter;
+	}
+
+	@Override
+	protected Query doToQuery(PredicateRequestContext context) {
+		if ( vector instanceof byte[] ) {
+			return new KnnByteVectorQuery(
+					absoluteFieldPath, (byte[]) vector, k, filter == null ? null : filter.toQuery( context ) );
+		}
+		if ( vector instanceof float[] ) {
+			return new KnnFloatVectorQuery(
+					absoluteFieldPath, (float[]) vector, k, filter == null ? null : filter.toQuery( context ) );
+		}
+
+		throw new UnsupportedOperationException(
+				"Unknown vector type " + vector.getClass() + ". only byte[] and float[] vectors are supported." );
+	}
+
+	public static class DefaultFactory<F>
+			extends AbstractLuceneValueFieldSearchQueryElementFactory<KnnPredicateBuilder, F> {
+		@Override
+		public KnnPredicateBuilder create(LuceneSearchIndexScope<?> scope, LuceneSearchIndexValueFieldContext<F> field) {
+			return new Builder<>( scope, field );
+		}
+	}
+
+	private static class Builder<F> extends AbstractBuilder implements KnnPredicateBuilder {
+		private final Class<?> vectorElementsType;
+		private int k;
+		private Object vector;
+		private LuceneSearchPredicate filter;
+
+		private Builder(LuceneSearchIndexScope<?> scope, LuceneSearchIndexValueFieldContext<F> field) {
+			super( scope, field );
+
+			LuceneFieldCodec<F> codec = field.type().codec();
+			if ( codec instanceof LuceneVectorFieldCodec ) {
+				vectorElementsType = ( (LuceneVectorFieldCodec<F>) codec ).vectorElementsType();
+			}
+			else {
+				// shouldn't really happen as if someone tries this it should fail on `queryElementFactory` lookup.
+				throw new IllegalStateException( "Attempting to use a knn predicate on a non-vector field." );
+			}
+		}
+
+		@Override
+		public void k(int k) {
+			this.k = k;
+		}
+
+		@Override
+		public void vector(Object vector) {
+			if ( !vector.getClass().isArray() ) {
+				throw new IllegalArgumentException( "Vector can only be either a float or a byte array (float[], byte[])." );
+			}
+			if ( !vectorElementsType.equals( vector.getClass().getComponentType() ) ) {
+				throw log.vectorKnnMatchVectorTypeDiffersFromField( absoluteFieldPath, vectorElementsType,
+						vector.getClass().getComponentType() );
+			}
+			this.vector = vector;
+		}
+
+		@Override
+		public void filter(SearchPredicate filter) {
+			this.filter = LuceneSearchPredicate.from( scope, filter );
+		}
+
+		@Override
+		public void boost(float boost) {
+			throw new UnsupportedOperationException( "This predicate does not support boost" );
```

Comment:
I wonder why you don't support boosts on KNN predicates... seems to me it's absolutely necessary that the KNN predicate returns some sort of score if we want to mix KNN with full-text search and use relevance sort... I think that was something that an OpenSearch hinted at: don't use KNN alone, but use it to improve the relevance of your full-text search?

Reply:
I wasn't sure if the mix will work out and that only a knn predicate at the root will work ok. So I though I'll start without boosts and then see where it takes us 😃

Reply:
Oh ok. Well starting small is perfectly fine. Just make sure you keep a list of the shortcuts you take... somewhere... so that we can add them to the backlog.